eva/optimizer/operators.py

```diff
@@ -621,6 +625,123 @@ def __hash__(self) -> int:
                      self.video_blob))
 
 
+class LogicalFunctionScan(Operator):
+    """
+    Logical node for function table scans
+
+    Attributes:
+        func_expr: AbstractExpression
+            function_expression that yield a table like output
+    """
+
+    def __init__(self,
+                 func_expr: AbstractExpression,
+                 children: List = None):
+        super().__init__(OperatorType.LOGICALFUNCTIONSCAN, children)
+        self._func_expr = func_expr
+
+    @property
+    def func_expr(self):
+        return self._func_expr
+
+    def __eq__(self, other):
+        is_subtree_equal = super().__eq__(other)
+        if not isinstance(other, LogicalFunctionScan):
+            return False
+        return (is_subtree_equal
+                and self.func_expr == other.func_expr)
+
+    def __hash__(self) -> int:
+        return hash((super().__hash__(),
+                     self.func_expr))
+
+
+class LogicalJoin(Operator):
+    """
+    Logical node for join operators
+
+    Attributes:
+        join_type: JoinType
+            Join type provided by the user - Lateral, Inner, Outer
+        join_predicate: AbstractExpression
+            condition/predicate expression used to join the tables
+    """
+
+    def __init__(self,
+                 join_type: JoinType,
+                 join_predicate: AbstractExpression = None,
```

Comment:
Shouldn't we have a LogicalPredicate Operator if we have a plan and executor?

Reply:
Same for project I guess.

Reply:
Do you mean the type of join_predicate should be LogicalPredicate?
If yes, then no. Right now we don't have the rule to push predicates into join so the optimised plan will always have a PredicateNode above join. I kept join_predicate in case we plan to push them later.